pyuvdata/telescopes.py

```diff
@@ -94,212 +149,426 @@ def __init__(self):
+            moon_loc.ellipsoid = self._location.ellipsoid
+            return moon_loc
+
+    @location_obj.setter
+    def location_obj(self, val):
+        if isinstance(val, EarthLocation):
+            self._location.frame = "itrs"
+        elif uvutils.hasmoon and isinstance(val, uvutils.MoonLocation):
+            self._location.frame = "mcmf"
+            self._location.ellipsoid = val.ellipsoid
+        else:
+            raise ValueError(
+                "location_obj is not a recognized location object. Must be an "
+                "EarthLocation or MoonLocation object."
+            )
+        self.location = np.array(
+            [val.x.to("m").value, val.y.to("m").value, val.z.to("m").value]
         )
 
-    elif telescope_name.lower() in telescope_list:
-        telescope_index = telescope_list.index(telescope_name.lower())
-        telescope_dict = telescope_dict_in[telescope_keys[telescope_index]]
-        obj = Telescope()
-        obj.citation = telescope_dict["citation"]
-        obj.telescope_name = telescope_keys[telescope_index]
-        if telescope_dict["center_xyz"] is not None:
-            obj.telescope_location = telescope_dict["center_xyz"]
-        else:
-            if (
-                telescope_dict["latitude"] is None
-                or telescope_dict["longitude"] is None
-                or telescope_dict["altitude"] is None
-            ):
+    def _set_uvcal_requirements(self):
+        """Set the UVParameter required fields appropriately for UVCal."""
+        self._name.required = True
+        self._location.required = True
+        self._instrument.required = False
+        self._Nants.required = True
+        self._antenna_names.required = True
+        self._antenna_numbers.required = True
+        self._antenna_positions.required = True
+        self._antenna_diameters.required = False
+        self._x_orientation.required = True
+
+    def _set_uvdata_requirements(self):
+        """Set the UVParameter required fields appropriately for UVData."""
+        self._name.required = True
+        self._location.required = True
+        self._instrument.required = True
+        self._Nants.required = True
+        self._antenna_names.required = True
+        self._antenna_numbers.required = True
+        self._antenna_positions.required = True
+        self._x_orientation.required = False
+        self._antenna_diameters.required = False
+
+    def _set_uvflag_requirements(self):
+        """Set the UVParameter required fields appropriately for UVCal."""
+        self._name.required = True
+        self._location.required = True
+        self._instrument.required = False
+        self._Nants.required = True
+        self._antenna_names.required = True
+        self._antenna_numbers.required = True
+        self._antenna_positions.required = True
+        self._antenna_diameters.required = False
+        self._x_orientation.required = False
+
+    def update_params_from_known_telescopes(
```

Comment:
This is used extensively by the other objects and I consolidated a bunch of code from the other objects here. This is because we have older UVCal and UVFlag files which have some of the telescope metadata but not all of it that is now required. Since we want to be able to read older files without erroring we need this code somewhere and I think it is much cleaner to be here rather than on the top level objects because here it can be consolidated.